File: Algorithms/Sorting/Bubble Sort/bubble.cpp
#include <iostream>
using namespace std;

void print_items(int *arr, int size)
{
    // print elements of the array
    for (int i = 0; i < size; i++)
    {
        cout << arr[i] << endl;
    }
}

void ascending(int *arr, int size)
{
    // outer loop to loop through passes
    for (int i = 0; i < size - 1; i++)
    {
        // inner loop to iterate over items
        bool flag = false;
        for (int j = 0; j < size - i - 1; j++)
        {
            // check if the items are not in ascending order
            if (arr[j] > arr[j + 1])
            {
                // perform swap operation
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                flag = true;
            }
        }
        // check if flag is false, meaning the array is already sorted
        if (flag == false)
        {
            break;
        }
    }
}

int main()
{
    cout << "_______-----------------------____________" << endl;
    cout << "-------------BUBBLE SORT---------------" << endl;

    int *array;
    int size;

    cout << "Enter the number of elements in the array: ";
    cin >> size;

    // allocate memory for the array
    array = new int[size];

    if (array == NULL)
    {
        cout << "Sorry, could not allocate memory for the array." << endl;
        exit(1);
    }

    cout << "Memory allocation successful!" << endl;

    // use allocated memory
    for (int i = 0; i < size; i++)
    {
        cout << "Enter item: " << endl;
        cin >> array[i];
    }
    cout << "Unsorted order" << endl;
    print_items(array, size);

    ascending(array, size);
    cout << "Ascending order" << endl;
    print_items(array, size);
    // free memory
    delete[] array;

    return 0;
}

File: Algorithms/Sorting/Bubble Sort/bubble.py
def bubble_sort(array):
    """Sorts the given array using the bubble sort algorithm."""
    n = len(array)
    for i in range(n):
        # Flag to check if any swaps were made in the current iteration
        swapped = False
        for j in range(0, n - i - 1):
            if array[j] > array[j + 1]:
                # Swap elements using tuple unpacking
                array[j], array[j + 1] = array[j + 1], array[j]
                swapped = True
        if not swapped:
            # If no swaps were made in the current iteration, the array is already sorted
            break

def display_items(array):
    """Displays the items in the array."""
    for i, item in enumerate(array):
        print(f"Item {i} is {item}")

array = [12, 14, 3, 5, 76, 78]

print("Items before sorting:")
display_items(array)
bubble_sort(array)
print("Items after sorting:")
display_items(array)
